src/EditorFeatures/CSharpTest/CodeActions/SyncNamespace/SyncNamespaceTests_ChangeNamespace.cs

```diff
@@ -0,0 +1,1858 @@
+        void {declaredNamespace}.Interface1.M1(Interface1 c1){{}}
+    }}
+}}</Document>
+    </Project>
+</Workspace>";
+
+            var expectedSourceOriginal =
+@"namespace A.B.C
+{
+    interface Interface1
+    {
+        void M1(Interface1 c1);
+    }
+}";
+            var expectedSourceReference =
+@"
+namespace Foo
+{
+    using A.B.C;
+
+    class RefClass : Interface1
+    {
+        void Interface1.M1(Interface1 c1){}
+    }
+}";
+            await TestChangeNamespaceAsync(code, expectedSourceOriginal, expectedSourceReference);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task ChangeNamespace_ChangeUsingsInMultipleContainers()
+        {
+            var defaultNamespace = "A";
+            var declaredNamespace = "Foo.Bar.Baz";
+
+            var documentPath1 = CreateDocumentFilePath(new[] { "B", "C" }, "File1.cs");
+            var documentPath2 = CreateDocumentFilePath(Array.Empty<string>(), "File2.cs");
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath1.folder}"" FilePath=""{documentPath1.filePath}""> 
+namespace [||]{declaredNamespace}
+{{
+    class Class1
+    {{
+    }}
+}}</Document>
+<Document Folders=""{documentPath2.folder}"" FilePath=""{documentPath2.filePath}""> 
+namespace NS1
+{{
+    using Foo.Bar.Baz;
+
+    class Class2
+    {{
+        Class1 c2;
+    }}
+
+    namespace NS2
+    {{
+        using Foo.Bar.Baz;
+
+        class Class2
+        {{
+            Class1 c1;
+        }}
+    }}
+}}</Document>
+    </Project>
+</Workspace>";
+
+            var expectedSourceOriginal =
+@"namespace A.B.C
+{
+    class Class1
+    {
+    }
+}";
+            var expectedSourceReference =
+@"
+namespace NS1
+{
+    using A.B.C;
+
+    class Class2
+    {
+        Class1 c2;
+    }
+
+    namespace NS2
+    {
+        class Class2
+        {
+            Class1 c1;
+        }
+    }
+}";
+            await TestChangeNamespaceAsync(code, expectedSourceOriginal, expectedSourceReference);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task ChangeNamespace_WithAliasReferencesInOtherDocument()
+        {
+            var defaultNamespace = "A";
+            var declaredNamespace = "Foo.Bar.Baz";
+
+            var documentPath1 = CreateDocumentFilePath(new[] { "B", "C" }, "File1.cs");
+            var documentPath2 = CreateDocumentFilePath(Array.Empty<string>(), "File2.cs");
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath1.folder}"" FilePath=""{documentPath1.filePath}""> 
+namespace [||]{declaredNamespace}
+{{
```

Comment:
a lot of these tests doesn't looks like sync namespace test but change namespace tests. for example, the test contains side effect of remove usings or simplifier but tests doesn't try to test all possible case of the effect of remove usings and simplifiers. but looks like sync namespace tests tries to test all possible cases for change namespace refactoring you created as a part of this feature.

Reply:
You are right, most of those tests just go a little bit further than simply testing "change namespace" (which is not an individual function yet), the only extra work is that the folder set in the XML decides the target namespace. So potentially I can refactor them later (once "change namespace" is done) to separate each test into two, first test constructing target namespace from folder, 2nd test the resulting source change given the target namespace.

Reply:
Whatever you do, amke sure the test reflects the experience the *user* has with the feature.  So if the input to the test is the code prior to the user action, and the specific action the user takes.  THe output should be hte final effect on the solution.

We don't want to test intermediary states, or subcomponents, as that will make a brittle dependency between the test and the implementation of the feature.  Instead, we want the test to describe the expected behavior of the feature no matter how it is implemented.  That way we can cahnge the impl as much as we want, without having to change any tests, and the tests will continue validating that we're getting the same net effect.